src/Shared/PropertyFetcher.AOT.cs

```diff
@@ -0,0 +1,237 @@
+
+    /// <summary>
+    /// Try to fetch the property from the object.
+    /// </summary>
+    /// <param name="obj">Object to be fetched.</param>
+    /// <param name="value">Fetched value.</param>
+    /// <returns><see langword= "true"/> if the property was fetched.</returns>
+#if NET6_0_OR_GREATER
+    [RequiresUnreferencedCode(TrimCompatibilityMessage)]
+#endif
+    public bool TryFetch(
+#if NETSTANDARD2_1_0_OR_GREATER || NET6_0_OR_GREATER
+        [NotNullWhen(true)]
+#endif
+        object? obj,
+        out T? value)
+    {
+        var innerFetcher = this.innerFetcher;
+        if (innerFetcher is null)
+        {
+            return TryFetchRare(obj, this.propertyName, ref this.innerFetcher, out value);
+        }
+
+        return innerFetcher.TryFetch(obj, out value);
+    }
+
+#if NET6_0_OR_GREATER
+    [RequiresUnreferencedCode(TrimCompatibilityMessage)]
+#endif
+    private static bool TryFetchRare(object? obj, string propertyName, ref PropertyFetch? destination, out T? value)
+    {
+        if (obj is null)
+        {
+            value = default;
+            return false;
+        }
+
+        var fetcher = PropertyFetch.Create(obj.GetType().GetTypeInfo(), propertyName);
+
+        if (fetcher is null)
+        {
+            value = default;
+            return false;
+        }
+
+        destination = fetcher;
+
+        return fetcher.TryFetch(obj, out value);
+    }
+
+    // see https://github.com/dotnet/corefx/blob/master/src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs
+#if NET6_0_OR_GREATER
+    [RequiresUnreferencedCode(TrimCompatibilityMessage)]
+#endif
+    private abstract class PropertyFetch
+    {
+        public abstract int NumberOfInnerFetchers { get; }
+
+        public static PropertyFetch? Create(TypeInfo type, string propertyName)
+        {
+            var property = type.DeclaredProperties.FirstOrDefault(p => string.Equals(p.Name, propertyName, StringComparison.OrdinalIgnoreCase)) ?? type.GetProperty(propertyName);
+            return CreateFetcherForProperty(property);
+
+            static PropertyFetch? CreateFetcherForProperty(PropertyInfo? propertyInfo)
+            {
+                if (propertyInfo == null || !typeof(T).IsAssignableFrom(propertyInfo.PropertyType))
+                {
+                    // returns null and wait for a valid payload to arrive.
+                    return null;
+                }
+
+                var declaringType = propertyInfo.DeclaringType;
+                if (declaringType!.IsValueType)
+                {
+                    throw new NotSupportedException(
+                        $"Type: {declaringType.FullName} is a value type. PropertyFetcher can only operate on reference payload types.");
+                }
+
+                if (declaringType == typeof(object))
+                {
+                    // TODO: REMOVE this if branch when .NET 7 is out of support.
```

Comment:
net7.0 is not one of the target for the project currently:
https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1415/files#:~:text=%3CTargetFrameworks%3Enet6.0%3Bnetstandard2.0%3Bnet462%3C/TargetFrameworks%3E
Do we need this here?

Reply:
I would like this to be a straight copy from opentelemetry-dotnet. Then we can keep the 2 in sync over time.

Also, the reason this code is here isn't in case one of the libraries targets net7.0, its for when the application targets net7.0, which we can't control here.